src/components/ControlBar.tsx

```diff
@@ -40,23 +43,75 @@ export const RestoreBtn = styled(Button)`
 `;
 
 interface ControlBarProps {
+  modelSaveSubs: Array<(save: ModelSave) => void>;
   worker: Worker;
 }
 
 export default function ControlBar(props: ControlBarProps): React.ReactElement {
-  const { worker } = props;
+  const { modelSaveSubs, worker } = props;
+
+  useEffect(() => {
+    modelSaveSubs.push((sav: ModelSave) => {
+      save(sav);
+      console.log('wrote a save');
+    });
+
+    // take the json and have the user download it
+    function save(sav: ModelSave): void {
+      const filename = `${sav.modelType}@${sav.time}.CFSS`;
+      const dat = JSON.stringify(sav);
+      const blob = new Blob([dat], { type: 'application/json' });
+      const url = URL.createObjectURL(blob);
+
+      const link = document.createElement('a');
+      link.href = url;
+      link.download = filename;
+      link.style.display = 'none';
+      document.body.appendChild(link);
+
+      link.click();
+
+      URL.revokeObjectURL(url);
+      document.body.removeChild(link);
+
+      console.log('wrote a save to ' + filename, sav);
+    }
+  }, [modelSaveSubs]);
+
+  // take a file and send its contents to the worker
+  function load(file: File): void {
+    console.log('reading file ', file);
+
+    const reader = new FileReader();
+    reader.onload = () => {
+      const text: string = reader.result?.toString() ?? '';
+      const data = JSON.parse(text) as ModelSave;
+      console.log('got', data);
+      worker.postMessage({ func: 'deserialize', args: data });
+    };
+    reader.readAsText(file);
+  }
+
   return (
     <>
       <SaveBtn
         onClick={() => {
-          console.log('Click Save');
+          worker.postMessage({ func: 'serialize' });
         }}
       >
         Save Model
       </SaveBtn>
       <RestoreBtn
         onClick={() => {
-          console.log('Click Restore');
+          fileDialog()
+            .then((files) => {
+              if (files.length > 0) {
+                load(files[0]);
+              }
+            })
+            .catch((e) => {
+              throw e;
+            });
```

Comment:
The model should be firstly stored in the `local storage`, cause we can not push users to download a file every time when they save the model. Downloading an independent file every time is not a good choice.

**localStorage** *is a web storage object that enables developers to store key-value pairs in a web browser and ensures that this data survives all page refreshes, even when a user closes or restarts a browser*

Reply:
https://blog.logrocket.com/localstorage-javascript-complete-guide/

Reply:
Or IndexDB, should probably implement a custom component to allow users select source first (PC/URL/LocalStorage) and create a List for LocalStorage if selected.

Reply:
sorry I'm not very familiar with browser IO and just went for the first option I saw. 

Given the size of the model output I think IndexDB might be a better choice, though I'd like if we could implement this in a separate PR.